gz-transport: Messages

## Configuration

* os: Ubuntu Noble
* transport: zenoh


## Links

* [Live](https://gazebosim.org/api/transport/15/messages.html)
* [Source](https://github.com/gazebosim/gz-transport/tree/gz-transport15/tutorials/04_messages.md)


## Process

    
            
### Step 1 - Instructions work
            
        

    
If the tutorial is an .sdf file, please open the file and read the comments at the top of the file.
        

To test with Zenoh. In each terminal, run `export GZ_TRANSPORT_IMPLEMENTATION=zenoh` before running any of the commands in the tutorial.

    

    
    
            
### Step 2 - Images (if there are any) match the result
            
        

    
If the tutorial is an .sdf file, post a screenshot of the simulation run.
        

    

    
    
            
### Step 3 - Comment status
            
        

    
Include a special comment at the end of your post to indicate pass/fail. Use either `[status: passed]` or `[status: failed]`.

[ source installed, Noble 24.04]
Below image is for  Publisher.cc and subscriber.cc
<img width="2746" height="1440" alt="Image" src="https://github.com/user-attachments/assets/1bb91c6b-ea7f-47c3-aabf-6f6b1bdde43e" />
Below one is for  Publisher.cc and subscriber_generic.cc file, here i changed generic file by adding a header then it worked
`#include <google/protobuf/text_format.h>
`
else i got error 
`make subscriber_generic
[ 50%] Building CXX object CMakeFiles/subscriber_generic.dir/subscriber_generic.cc.o
/home/sam/gz_transport_tutorial/subscriber_generic.cc: In function ‘void cb(const google::protobuf::Message&, const gz::transport::v15::MessageInfo&)’:
/home/sam/gz_transport_tutorial/subscriber_generic.cc:11:42: error: ‘google::protobuf::TextFormat’ has not been declared
   11 |   if (std::string str; google::protobuf::TextFormat::PrintToString(_msg, &str))
      |                                          ^~~~~~~~~~
make[3]: *** [CMakeFiles/subscriber_generic.dir/build.make:76: CMakeFiles/subscriber_generic.dir/subscriber_generic.cc.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:139: CMakeFiles/subscriber_generic.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:146: CMakeFiles/subscriber_generic.dir/rule] Error 2
make: *** [Makefile:150: subscriber_generic] Error 2
`
<img width="2746" height="1440" alt="Image" src="https://github.com/user-attachments/assets/9d53dd77-94ac-427e-aebc-f56415726439" />

Below is for "Using custom Protobuf":

<img width="2746" height="1440" alt="Image" src="https://github.com/user-attachments/assets/5148cb04-b7de-4133-a7aa-9b60037f05c1" />

Below for topic remaping:

<img width="2746" height="1440" alt="Image" src="https://github.com/user-attachments/assets/eefcb412-9971-4cc4-a044-cb5765448514" />

[status: passed]

[ source installed, Noble 24.04] {zenoh}

Below image is for Publisher.cc and subscriber.cc

<img width="2746" height="1703" alt="Image" src="https://github.com/user-attachments/assets/2607eb7f-76d9-412c-a287-390555d8abd3" />

Below one is for Publisher.cc and subscriber_generic.cc file, here i changed generic file by adding a header then it worked
else i got error 
`make subscriber_generic [ 50%] Building CXX object CMakeFiles/subscriber_generic.dir/subscriber_generic.cc.o /home/sam/gz_transport_tutorial/subscriber_generic.cc: In function ‘void cb(const google::protobuf::Message&, const gz::transport::v15::MessageInfo&)’: /home/sam/gz_transport_tutorial/subscriber_generic.cc:11:42: error: ‘google::protobuf::TextFormat’ has not been declared 11 |   if (std::string str; google::protobuf::TextFormat::PrintToString(_msg, &str)) |                                          ^~~~~~~~~~ make[3]: *** [CMakeFiles/subscriber_generic.dir/build.make:76: CMakeFiles/subscriber_generic.dir/subscriber_generic.cc.o] Error 1 make[2]: *** [CMakeFiles/Makefile2:139: CMakeFiles/subscriber_generic.dir/all] Error 2 make[1]: *** [CMakeFiles/Makefile2:146: CMakeFiles/subscriber_generic.dir/rule] Error 2 make: *** [Makefile:150: subscriber_generic] Error 2 
Image`

Below is for "Using custom Protobuf":

<img width="2746" height="1703" alt="Image" src="https://github.com/user-attachments/assets/2e38bcaf-d185-457b-998c-84a8773565fa" />

Below for topic remaping:

<img width="2746" height="1703" alt="Image" src="https://github.com/user-attachments/assets/5dc73a08-eb6f-4f91-bc18-4f96336bb9bf" />

<img width="2746" height="1703" alt="Image" src="https://github.com/user-attachments/assets/4983c467-1dee-44dc-8db2-2af28ca6486a" />

--- a/src/SubscriptionHandler.cc
+++ b/src/SubscriptionHandler.cc
@@ -175,7 +175,8 @@ namespace gz::transport
 
         auto output = this->CreateMsg(
           _sample.get_payload().as_string(), msgType);
-        this->RunLocalCallback(*output, msgInfo);
+        if (output)
+          this->RunLocalCallback(*output, msgInfo);
       }
       else
       {

